In [32]:
import pandas as pd

In [82]:
weather_df = pd.read_csv("london_2017_weather_data_hourly&daily.csv")

weather_df.describe()

temperature_2m (°C)  relative_humidity_2m (%)  \
count          8760.000000               8760.000000   
mean             11.200400                 79.571918   
std               6.044539                 13.581923   
min              -4.800000                 36.000000   
25%               6.900000                 71.000000   
50%              11.400000                 83.000000   
75%              15.500000                 90.000000   
max              31.600000                100.000000   

       apparent_temperature (°C)  precipitation (mm)    rain (mm)  \
count                8760.000000         8760.000000  8760.000000   
mean                    8.610902            0.075320     0.072660   
std                     7.135921            0.291952     0.276574   
min                    -9.100000            0.000000     0.000000   
25%                     3.500000            0.000000     0.000000   
50%                     8.500000            0.000000     0.000000   
75%                    13.900000            0.000000     0.000000   
max                    33.500000            4.700000     4.700000   

       snowfall (cm)  weather_code (wmo code)  cloud_cover (%)  \
count    8760.000000              8760.000000      8760.000000   
mean        0.001862                 9.923402        59.715068   
std         0.047721                18.971835        37.132761   
min         0.000000                 0.000000         0.000000   
25%         0.000000                 1.000000        26.000000   
50%         0.000000                 2.000000        66.000000   
75%         0.000000                 3.000000       100.000000   
max         2.380000                75.000000       100.000000   

       wind_speed_100m (km/h)  
count             8760.000000  
mean                24.521587  
std                 10.741635  
min                  0.500000  
25%                 17.100000  
50%                 23.800000  
75%                 31.100000  
max                 77.500000

we have 8760 which is equivalent to the number of days in a year

In [83]:
weather_df.rename(columns={"temperature_2m (°C)":"real_temp",
                          "apparent_temperature (°C)":"feel_temp",
                           "relative_humidity_2m (%)": "humidity",
                           "precipitation (mm)":"precipitation",
                           "rain (mm)":"rain",
                           "snowfall (cm)":"snowfall",
                           "weather_code (wmo code)":"weather_code",
                           "cloud_cover (%)":"cloud_cover",
                           "wind_speed_100m (km/h)":"wind_speed"
                          }, inplace=True)


In [24]:
#weather_df.set_index("time")

Weather DF is ready

In [84]:
trip_df = pd.read_csv("clean_london_all_2017_trip_data.csv")
trip_df.rename(columns={"Unnamed: 0":"index"}, inplace=True)
trip_df.set_index("index", inplace=True)

trip_df

Rental Id  Duration  Bike Id          End Date  EndStation Id  \
index                                                                     
0          62747748     420.0   4347.0  01/03/2017 21:08          295.0   
1          62746620     300.0   2911.0  01/03/2017 19:47          184.0   
2          62727993     420.0  10963.0  01/03/2017 08:37          104.0   
3          62743379     420.0  10490.0  01/03/2017 18:16          173.0   
4          62733782     960.0  10864.0  01/03/2017 11:39          257.0   
...             ...       ...      ...               ...            ...   
10361683   70965745     480.0   8477.0  31/10/2017 23:57          134.0   
10361684   70965743     420.0   6149.0  31/10/2017 23:56          387.0   
10361685   70965748     180.0  10524.0  31/10/2017 23:53          345.0   
10361686   70965762     300.0   5676.0  31/10/2017 23:58          707.0   
10361687   70965763     240.0   4570.0  31/10/2017 23:58          707.0   

                                EndStation Name           Start Date  \
index                                                                  
0                      Swan Street, The Borough  2017-03-01 21:00:01   
1                    Portland Place, Marylebone  2017-03-01 19:00:42   
2                              Crosswall, Tower  2017-03-01 08:00:30   
3                     Waterloo Road, South Bank  2017-03-01 18:00:09   
4            Westminster University, Marylebone  2017-03-01 11:00:23   
...                                         ...                  ...   
10361683           Wapping High Street, Wapping  2017-10-31 23:49:00   
10361684            Fire Brigade Pier, Vauxhall  2017-10-31 23:49:00   
10361685                  Flood Street, Chelsea  2017-10-31 23:50:00   
10361686  Barons Court Station, West Kensington  2017-10-31 23:53:00   
10361687  Barons Court Station, West Kensington  2017-10-31 23:54:00   

          StartStation Id                    StartStation Name  
index                                                           
0                     803           Southwark Street, Bankside  
1                     257   Westminster University, Marylebone  
2                     101                 Queen Street 1, Bank  
3                       9             New Globe Walk, Bankside  
4                     238          Frampton Street, Paddington  
...                   ...                                  ...  
10361683              251  Brushfield Street, Liverpool Street  
10361684              360            Howick Place, Westminster  
10361685              430                South Parade, Chelsea  
10361686              635          Greyhound Road, Hammersmith  
10361687              635          Greyhound Road, Hammersmith  

[10337102 rows x 9 columns]

In [85]:
trip_df.sort_values(by=['Start Date'], inplace=True)
trip_df

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6688701   61395912    2460.0   6270.0  2017-01-01 00:43:00          368.0   
6688709   61395913     960.0  13994.0  2017-01-01 00:18:00          256.0   
6688708   61395914    2820.0   1110.0  2017-01-01 00:49:00          233.0   
6688707   61395915    2460.0   4555.0  2017-01-01 00:43:00          368.0   
6688702   61395917    6180.0   2280.0  2017-01-01 01:45:00           64.0   
...            ...       ...      ...                  ...            ...   
6770971   72337677    1320.0   8903.0  2018-01-01 00:21:00          228.0   
6770972   72337671    2820.0   9562.0  2018-01-01 00:46:00          333.0   
6770973   72337667    2460.0  10830.0  2018-01-01 00:40:00           39.0   
6770974   72337668     300.0   9833.0  2018-01-01 00:04:00          114.0   
6770975   72337669     480.0   2879.0  2018-01-01 00:07:00          376.0   

                                     EndStation Name           Start Date  \
index                                                                       
6688701                Harriet Street, Knightsbridge  2017-01-01 00:02:00   
6688709                      Houghton Street, Strand  2017-01-01 00:02:00   
6688708                     Pall Mall East, West End  2017-01-01 00:02:00   
6688707                Harriet Street, Knightsbridge  2017-01-01 00:02:00   
6688702                    William IV Street, Strand  2017-01-01 00:02:00   
...                                              ...                  ...   
6770971              St. James's Square, St. James's  2017-12-31 23:59:00   
6770972         Palace Gardens Terrace, Notting Hill  2017-12-31 23:59:00   
6770973           Shoreditch High Street, Shoreditch  2017-12-31 23:59:00   
6770974  Park Road (Baker Street), The Regent's Park  2017-12-31 23:59:00   
6770975                      Millbank Tower, Pimlico  2017-12-31 23:59:00   

         StartStation Id                     StartStation Name  
index                                                           
6688701              368         Harriet Street, Knightsbridge  
6688709              170          Hardwick Street, Clerkenwell  
6688708              105           Westbourne Grove, Bayswater  
6688707              368         Harriet Street, Knightsbridge  
6688702              687                 Maclise Road, Olympia  
...                  ...                                   ...  
6770971              213            Wellington Arch, Hyde Park  
6770972              333  Palace Gardens Terrace, Notting Hill  
6770973              132        Bethnal Green Road, Shoreditch  
6770974                7     Charlbert Street, St. John's Wood  
6770975              245               Grosvenor Road, Pimlico  

[10337102 rows x 9 columns]

In [86]:
import numpy as np
#removing the unnamed:0 
trip_df[trip_df['Rental Id'] == 61416022]

#trip_df['Start Date Clean'] = np.where(pd.to_datetime(trip_df['Start Date'].str.contains('-'), format= '%Y-%m-%d %H:%M:%S')
#    ,trip_df['Start Date'].str.contains('/'), pd.to_datetime(trip_df['Start Date'], format='%d/%m/%Y %H:%M')
#                                  )

#trip_df['Start Date Clean'] = np.where(trip_df['Start Date'].str.contains('-'),  pd.to_datetime(trip_df['Start Date'], format='%d/%m/%Y %H:%M').pd.to_datetime(trip_df['Start Date'], format= '%Y-%m-%d %H:%M:%S'),)

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6707469   61416022    3600.0   3474.0  2017-01-03 07:42:00          510.0   

                  EndStation Name           Start Date  StartStation Id  \
index                                                                     
6707469  Westferry DLR, Limehouse  2017-01-03 06:42:00              201   

                 StartStation Name  
index                               
6707469  Dorset Square, Marylebone

In [38]:
trip_df

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6688701   61395912    2460.0   6270.0  2017-01-01 00:43:00          368.0   
6688709   61395913     960.0  13994.0  2017-01-01 00:18:00          256.0   
6688708   61395914    2820.0   1110.0  2017-01-01 00:49:00          233.0   
6688707   61395915    2460.0   4555.0  2017-01-01 00:43:00          368.0   
6688702   61395917    6180.0   2280.0  2017-01-01 01:45:00           64.0   
...            ...       ...      ...                  ...            ...   
6770971   72337677    1320.0   8903.0  2018-01-01 00:21:00          228.0   
6770972   72337671    2820.0   9562.0  2018-01-01 00:46:00          333.0   
6770973   72337667    2460.0  10830.0  2018-01-01 00:40:00           39.0   
6770974   72337668     300.0   9833.0  2018-01-01 00:04:00          114.0   
6770975   72337669     480.0   2879.0  2018-01-01 00:07:00          376.0   

                                     EndStation Name           Start Date  \
index                                                                       
6688701                Harriet Street, Knightsbridge  2017-01-01 00:02:00   
6688709                      Houghton Street, Strand  2017-01-01 00:02:00   
6688708                     Pall Mall East, West End  2017-01-01 00:02:00   
6688707                Harriet Street, Knightsbridge  2017-01-01 00:02:00   
6688702                    William IV Street, Strand  2017-01-01 00:02:00   
...                                              ...                  ...   
6770971              St. James's Square, St. James's  2017-12-31 23:59:00   
6770972         Palace Gardens Terrace, Notting Hill  2017-12-31 23:59:00   
6770973           Shoreditch High Street, Shoreditch  2017-12-31 23:59:00   
6770974  Park Road (Baker Street), The Regent's Park  2017-12-31 23:59:00   
6770975                      Millbank Tower, Pimlico  2017-12-31 23:59:00   

         StartStation Id                     StartStation Name  
index                                                           
6688701              368         Harriet Street, Knightsbridge  
6688709              170          Hardwick Street, Clerkenwell  
6688708              105           Westbourne Grove, Bayswater  
6688707              368         Harriet Street, Knightsbridge  
6688702              687                 Maclise Road, Olympia  
...                  ...                                   ...  
6770971              213            Wellington Arch, Hyde Park  
6770972              333  Palace Gardens Terrace, Notting Hill  
6770973              132        Bethnal Green Road, Shoreditch  
6770974                7     Charlbert Street, St. John's Wood  
6770975              245               Grosvenor Road, Pimlico  

[10337102 rows x 9 columns]

I have inconsistent Date formats

In [3]:
#Working

trip_df['Start Date'] = pd.to_datetime(trip_df['Start Date'], dayfirst=True, format="mixed", errors='ignore') #'%d/%m/%Y %H:%M',
#trip_df['Start Date'] = pd.to_datetime(trip_df['Start Date'], yearfirst=True, exact=False)

weather_df['time'] = pd.to_datetime(weather_df['time'], format='%Y-%m-%dT%H:%M', errors='coerce')

NameError: name 'pd' is not defined

In [2]:
april_dates = trip_df[trip_df['Start Date'].dt.month == 2] #15 to 22
april_dates = april_dates[april_dates['Start Date'].dt.day == 1] #15 to 22
april_dates

NameError: name 'trip_df' is not defined

In [59]:
#This trip had inconsistent Data format 
trip_df[trip_df['Rental Id'] == 61416022]

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6707469   61416022    3600.0   3474.0  2017-01-03 07:42:00          510.0   

                  EndStation Name          Start Date  StartStation Id  \
index                                                                    
6707469  Westferry DLR, Limehouse 2017-01-03 06:42:00              201   

                 StartStation Name  
index                               
6707469  Dorset Square, Marylebone

In [60]:
trip_df

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6688701   61395912    2460.0   6270.0  2017-01-01 00:43:00          368.0   
6688709   61395913     960.0  13994.0  2017-01-01 00:18:00          256.0   
6688708   61395914    2820.0   1110.0  2017-01-01 00:49:00          233.0   
6688707   61395915    2460.0   4555.0  2017-01-01 00:43:00          368.0   
6688702   61395917    6180.0   2280.0  2017-01-01 01:45:00           64.0   
...            ...       ...      ...                  ...            ...   
6770971   72337677    1320.0   8903.0  2018-01-01 00:21:00          228.0   
6770972   72337671    2820.0   9562.0  2018-01-01 00:46:00          333.0   
6770973   72337667    2460.0  10830.0  2018-01-01 00:40:00           39.0   
6770974   72337668     300.0   9833.0  2018-01-01 00:04:00          114.0   
6770975   72337669     480.0   2879.0  2018-01-01 00:07:00          376.0   

                                     EndStation Name          Start Date  \
index                                                                      
6688701                Harriet Street, Knightsbridge 2017-01-01 00:02:00   
6688709                      Houghton Street, Strand 2017-01-01 00:02:00   
6688708                     Pall Mall East, West End 2017-01-01 00:02:00   
6688707                Harriet Street, Knightsbridge 2017-01-01 00:02:00   
6688702                    William IV Street, Strand 2017-01-01 00:02:00   
...                                              ...                 ...   
6770971              St. James's Square, St. James's 2017-12-31 23:59:00   
6770972         Palace Gardens Terrace, Notting Hill 2017-12-31 23:59:00   
6770973           Shoreditch High Street, Shoreditch 2017-12-31 23:59:00   
6770974  Park Road (Baker Street), The Regent's Park 2017-12-31 23:59:00   
6770975                      Millbank Tower, Pimlico 2017-12-31 23:59:00   

         StartStation Id                     StartStation Name  
index                                                           
6688701              368         Harriet Street, Knightsbridge  
6688709              170          Hardwick Street, Clerkenwell  
6688708              105           Westbourne Grove, Bayswater  
6688707              368         Harriet Street, Knightsbridge  
6688702              687                 Maclise Road, Olympia  
...                  ...                                   ...  
6770971              213            Wellington Arch, Hyde Park  
6770972              333  Palace Gardens Terrace, Notting Hill  
6770973              132        Bethnal Green Road, Shoreditch  
6770974                7     Charlbert Street, St. John's Wood  
6770975              245               Grosvenor Road, Pimlico  

[10337102 rows x 9 columns]

In [67]:
trip_df.sort_values(by=['Start Date'], inplace=True)
weather_df.sort_values(by=['time'], inplace=True)

In [69]:
trip_df

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
index                                                                       
6688701   61395912    2460.0   6270.0  2017-01-01 00:43:00          368.0   
6688709   61395913     960.0  13994.0  2017-01-01 00:18:00          256.0   
6688708   61395914    2820.0   1110.0  2017-01-01 00:49:00          233.0   
6688707   61395915    2460.0   4555.0  2017-01-01 00:43:00          368.0   
6688702   61395917    6180.0   2280.0  2017-01-01 01:45:00           64.0   
...            ...       ...      ...                  ...            ...   
6770973   72337667    2460.0  10830.0  2018-01-01 00:40:00           39.0   
6770974   72337668     300.0   9833.0  2018-01-01 00:04:00          114.0   
6770971   72337677    1320.0   8903.0  2018-01-01 00:21:00          228.0   
6770972   72337671    2820.0   9562.0  2018-01-01 00:46:00          333.0   
6770975   72337669     480.0   2879.0  2018-01-01 00:07:00          376.0   

                                     EndStation Name          Start Date  \
index                                                                      
6688701                Harriet Street, Knightsbridge 2017-01-01 00:02:00   
6688709                      Houghton Street, Strand 2017-01-01 00:02:00   
6688708                     Pall Mall East, West End 2017-01-01 00:02:00   
6688707                Harriet Street, Knightsbridge 2017-01-01 00:02:00   
6688702                    William IV Street, Strand 2017-01-01 00:02:00   
...                                              ...                 ...   
6770973           Shoreditch High Street, Shoreditch 2017-12-31 23:59:00   
6770974  Park Road (Baker Street), The Regent's Park 2017-12-31 23:59:00   
6770971              St. James's Square, St. James's 2017-12-31 23:59:00   
6770972         Palace Gardens Terrace, Notting Hill 2017-12-31 23:59:00   
6770975                      Millbank Tower, Pimlico 2017-12-31 23:59:00   

         StartStation Id                     StartStation Name  
index                                                           
6688701              368         Harriet Street, Knightsbridge  
6688709              170          Hardwick Street, Clerkenwell  
6688708              105           Westbourne Grove, Bayswater  
6688707              368         Harriet Street, Knightsbridge  
6688702              687                 Maclise Road, Olympia  
...                  ...                                   ...  
6770973              132        Bethnal Green Road, Shoreditch  
6770974                7     Charlbert Street, St. John's Wood  
6770971              213            Wellington Arch, Hyde Park  
6770972              333  Palace Gardens Terrace, Notting Hill  
6770975              245               Grosvenor Road, Pimlico  

[10337102 rows x 9 columns]

In [63]:
#trip_df["Start Date"].isna().any()
#trip_df[trip_df['Start Date'].isna()]

#There are some trips with NaN End Date
trip_df[trip_df['Rental Id'] == 70965788]

Empty DataFrame
Columns: [Rental Id, Duration, Bike Id, End Date, EndStation Id, EndStation Name, Start Date, StartStation Id, StartStation Name]
Index: []

In [70]:
#merging and keeping the weather time column
#merged_df = pd.merge_asof(trip_df, weather_df, direction='nearest', left_on='Start Date', right_on='time')

#merging on Start Date
merged_df = pd.merge_asof(trip_df, weather_df, direction='nearest', left_on='Start Date', right_on='time')

In [72]:
#Dropping the weather time column:

merged_df.drop(['time'], axis=1)

Rental Id  Duration  Bike Id             End Date  EndStation Id  \
0          61395912    2460.0   6270.0  2017-01-01 00:43:00          368.0   
1          61395913     960.0  13994.0  2017-01-01 00:18:00          256.0   
2          61395914    2820.0   1110.0  2017-01-01 00:49:00          233.0   
3          61395915    2460.0   4555.0  2017-01-01 00:43:00          368.0   
4          61395917    6180.0   2280.0  2017-01-01 01:45:00           64.0   
...             ...       ...      ...                  ...            ...   
10337097   72337667    2460.0  10830.0  2018-01-01 00:40:00           39.0   
10337098   72337668     300.0   9833.0  2018-01-01 00:04:00          114.0   
10337099   72337677    1320.0   8903.0  2018-01-01 00:21:00          228.0   
10337100   72337671    2820.0   9562.0  2018-01-01 00:46:00          333.0   
10337101   72337669     480.0   2879.0  2018-01-01 00:07:00          376.0   

                                      EndStation Name          Start Date  \
0                       Harriet Street, Knightsbridge 2017-01-01 00:02:00   
1                             Houghton Street, Strand 2017-01-01 00:02:00   
2                            Pall Mall East, West End 2017-01-01 00:02:00   
3                       Harriet Street, Knightsbridge 2017-01-01 00:02:00   
4                           William IV Street, Strand 2017-01-01 00:02:00   
...                                               ...                 ...   
10337097           Shoreditch High Street, Shoreditch 2017-12-31 23:59:00   
10337098  Park Road (Baker Street), The Regent's Park 2017-12-31 23:59:00   
10337099              St. James's Square, St. James's 2017-12-31 23:59:00   
10337100         Palace Gardens Terrace, Notting Hill 2017-12-31 23:59:00   
10337101                      Millbank Tower, Pimlico 2017-12-31 23:59:00   

          StartStation Id                     StartStation Name  real_temp  \
0                     368         Harriet Street, Knightsbridge        5.5   
1                     170          Hardwick Street, Clerkenwell        5.5   
2                     105           Westbourne Grove, Bayswater        5.5   
3                     368         Harriet Street, Knightsbridge        5.5   
4                     687                 Maclise Road, Olympia        5.5   
...                   ...                                   ...        ...   
10337097              132        Bethnal Green Road, Shoreditch        7.1   
10337098                7     Charlbert Street, St. John's Wood        7.1   
10337099              213            Wellington Arch, Hyde Park        7.1   
10337100              333  Palace Gardens Terrace, Notting Hill        7.1   
10337101              245               Grosvenor Road, Pimlico        7.1   

          humidity  feel_temp  precipitation  rain  snowfall  weather_code  \
0               96        2.3            0.0   0.0       0.0             2   
1               96        2.3            0.0   0.0       0.0             2   
2               96        2.3            0.0   0.0       0.0             2   
3               96        2.3            0.0   0.0       0.0             2   
4               96        2.3            0.0   0.0       0.0             2   
...            ...        ...            ...   ...       ...           ...   
10337097        77        1.0            0.0   0.0       0.0             2   
10337098        77        1.0            0.0   0.0       0.0             2   
10337099        77        1.0            0.0   0.0       0.0             2   
10337100        77        1.0            0.0   0.0       0.0             2   
10337101        77        1.0            0.0   0.0       0.0             2   

          cloud_cover  wind_speed  
0                  68        25.8  
1                  68        25.8  
2                  68        25.8  
3                  68        25.8  
4                  68        25.8  
...               ...         ...  
10337097           72 

In [1]:
april_dates = merged_df[merged_df['Start Date'].dt.month == 10] #15 to 22
april_dates = april_dates[april_dates['Start Date'].dt.day == 1] #15 to 22
april_dates

NameError: name 'merged_df' is not defined

In [81]:
april_dates.to_csv("march_first_data.csv", index=False)  

In [77]:
merged_df.dropna(how='any', inplace=True)

In [78]:
merged_df.isnull().sum()


Rental Id            0
Duration             0
Bike Id              0
End Date             0
EndStation Id        0
EndStation Name      0
Start Date           0
StartStation Id      0
StartStation Name    0
time                 0
real_temp            0
humidity             0
feel_temp            0
precipitation        0
rain                 0
snowfall             0
weather_code         0
cloud_cover          0
wind_speed           0
dtype: int64

In [79]:
merged_df.to_csv("london_2017_trip&weather_data.csv", index=False)  

https://stackoverflow.com/questions/40643565/inconsistent-datetime-format-in-pandas